docs/machine-learning/basic-concepts-model-training-in-mldotnet.MD

```diff
@@ -0,0 +1,169 @@
+
+An interesting property to note is that you can create a new data reader by 'attaching' a transformer to an existing data reader. This way you can have 'reader' with transformation behavior baked in:
+
+```c#
+var newReader = reader.Append(transformer1).Append(transformer2)
+```
+
+Another similarity to transformers is that, since data is lazily evaluated, *readers are lazy*: no (or minimal) actual 'reading' happens when you call `dataReader.Read()`: only when a cursor is requested on the resulting data does the reader begin to work.
+
+## Estimator
+
+The *estimator* is an object that learns from data. The result of the learning is a *transformer*.
+Here is the interface of `IEstimator<T>`:
+
+```c#
+public interface IEstimator<out TTransformer>
+    where TTransformer : ITransformer
+{
+    TTransformer Fit(IDataView input);
+    SchemaShape GetOutputSchema(SchemaShape inputSchema);
+}
+```
+
+You can easily imagine how *a sequence of estimators can be phrased as an estimator* of its own. In ML.NET, we rely on this property to create 'learning pipelines' that chain together different estimators:
+
+```c#
+var env = new LocalEnvironment(); // Initialize the ML.NET environment.
+var estimator = new ConcatEstimator(env, "Features", "SepalLength", "SepalWidth", "PetalLength", "PetalWidth")
+    .Append(new ToKeyEstimator(env, "Label"))
+    .Append(new SdcaMultiClassTrainer(env, "Features", "Label")) // This is the actual 'machine learning algorithm'.
+    .Append(new ToValueEstimator(env, "PredictedLabel"));
+
+var endToEndModel = estimator.Fit(data); // This now contains all the transformers that were used at training.
+```
+
+One important property of estimators is that *estimators are eager, not lazy*: every call to `Fit` is causing 'learning' to happen, which is potentially a time-consuming operation.
+
+## Prediction function
+
+The prediction function can be seen as a machine that applies a transformer to one 'row', such as at prediction time.
+
+Once we obtain the model (a *transformer* either trained via `Fit()`, or loaded from somewhere), we can use it to make 'predictions' using the normal calls to `model.Transform(data)`. However, when we use this model in a real life scenario, we often don't have a whole 'batch' of examples to predict on. Instead, we have one example at a time, and we need to make timely predictions on them immediately.
+
+Of course, we can reduce this to the batch prediction:
+
+- Create a data view with exactly one row.
+- Call `model.Transform(data)` to obtain the 'predicted data view'.
+- Get a cursor over the resulting data.
+- Advance the cursor one step to get to the first (and only) row.
+- Extract the predicted values out of it.
+
+The above algorithm can be implemented using the [schema comprehension](SchemaComprehension.md), with two user-defined objects `InputExample` and `OutputPrediction` as follows:
```

Comment:
This doc ref is for the Schema comprehension doc in the repo: https://github.com/dotnet/machinelearning/blob/master/docs/code/SchemaComprehension.md

Reply:
Good catch!